### PR TITLE
AVM1: Implement Error object (for #251 and #731)

### PR DESCRIPTION
**Pull request by Ace4896**
From the resources I've read, this looks correct, but I need some help with writing tests for the `toString` function (unsure about how the testing works in this project atm).

### CODE REVIEW

**Comment by Dinnerbone** [comment]
For testing, we generally make a Flash movie that asserts its state using `trace(foo)`. We take the output of that movie from Flash and store it in a file `output.txt`, and then when added to https://github.com/ruffle-rs/ruffle/blob/master/core/tests/regression_tests.rs it will automatically compare Ruffle's output of that swf to `output.txt`. Using this, we can assert that our behaviour matches Flash's behaviour.

We store these tests in [core/tests/swfs/avm1](https://github.com/ruffle-rs/ruffle/tree/master/core/tests/swfs/avm1). You can look at the existing ones for inspiration, but basically you're looking to try to use the thing you're testing (in this case, `new Error()`, `error.name = "foo"`, etc) in as many ways as possible - unusual ones too, to find edgecases. I imagine this class will be fairly quick and easy to test.

You'll need a Flash editor installed. That's unfortunately not trivial to get, these days. If you need help, you can ask me or someone on Discord to make the tests for you - or, you can write an actionscript test and have us turn that into a swf for you.

**Comment by Ace4896** [comment]
Thanks for the explanation @Dinnerbone, makes more sense now.

I wasn't able to setup a Flash IDE (currently trying to get FlashDevelop to work), so I couldn't run the tests I've written. For now though, do you mind making an swf with this?

```as
package
{
	import flash.display.Sprite;
	
	/**
	 * Testing for the "Error" object.
	 * @author Jon Pacheco
	 */
	public class Main extends Sprite 
	{
		public function Main() 
		{
			// Expected Output: "Error"
			var e = new Error();
			trace(e);
			
			// Expected Output: "TestError"
			e = new Error();
			e.name = "TestError";
			trace(e);
			
			// Expected Output: "Error: A"
			e = new Error("A");
			trace(e);
			
			// Expected Output: "Error: B"
			e = new Error();
			e.message = "B";
			trace(e);
			
			// Expected Output: "Error: 25"
			e = new Error(25);
			trace(e);
			
			// Expected Output: "TestError: A"
			e = new Error("A");
			e.name = "TestError";
			trace(e);
		}
	}
}
```

File: core/src/avm1/globals/error.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    object.force_set_function(
+        "toString",
+        to_string,
+        gc_context,
+        DontDelete | ReadOnly | DontEnum,
+        Some(fn_proto),
+    );
+
+    object.into()
```
Comment (Herschel) [inline core/src/avm1/globals/error.rs:44]: "message" and "name" should be properties on the prototype, defaulted to Error:
```
trace(Error.prototype.message); // "Error"
trace(Error.prototype.name); // "Error"
```

File: core/src/avm1/globals/error.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    let message = match args.get(0).cloned() {
+        Some(Value::String(s)) => s,
+        Some(v) => v.coerce_to_string(activation, context)?.to_string(),
+        _ => String::new(),
+    };
+
+    this.set("message", message.into(), activation, context)?;
+
+    // Default name for the object is "Error", which isn't provided by constructor
+    this.set("name", "Error".into(), activation, context)?;
```
Comment (Herschel) [inline core/src/avm1/globals/error.rs:24]: These defaults come from the prototype so shouldn't be set here.

File: core/src/avm1/globals/error.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    activation: &mut Activation<'_, 'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let message = match args.get(0).cloned() {
+        Some(Value::String(s)) => s,
+        Some(v) => v.coerce_to_string(activation, context)?.to_string(),
+        _ => String::new(),
+    };
```
Comment (Herschel) [inline core/src/avm1/globals/error.rs:19]: Looks like this should actually just copy the value in as is:
```
var e = new Error(111);
trace(e.message); // 111
trace(typeof e.message); // number
```
And it only does so if the arg is not undefined:
```
var e = new Error(undefined);
trace(e.message); // Error
trace(e.hasOwnProperty("message")); // false

var e = new Error("TEST");
trace(e.message); // TEST
trace(e.hasOwnProperty("message")); // true
```

File: core/src/avm1/globals/error.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    let message = this.get("message", activation, context)?;
+    let message_str = message.coerce_to_string(activation, context)?;
+
+    let formatted_message = if message_str.is_empty() {
+        name_str.to_string()
+    } else {
+        format!("{}: {}", name_str, message_str)
+    };
+
+    Ok(formatted_message.into())
```
Comment (Herschel) [inline core/src/avm1/globals/error.rs:65]: Looks like it only prints out the `message` for me:

```
var e = new Error();
e.name = "name";
e.message = "message";

trace(e.toString()); // message
```

**Comment by Herschel** [comment]
Thanks for the PR, and welcome! I posted some changes above -- mostly `name` and `message` seem to be dumb properties that the user can set to whatever. The test code you posted is actually AS3, you'll unfortunately still need to Flash IDE to write AS2 code. If you don't have easy access to it, I'd be happy to write up the test case for this PR.

File: core/src/avm1/globals/error.rs (hunk 2 above)
Comment (Ace4896) [reply core/src/avm1/globals/error.rs:24]: Ah, do you mean using `object.add_property` in `create_proto` instead?

File: core/src/avm1/globals/error.rs (hunk 4 above)
Comment (Ace4896) [reply core/src/avm1/globals/error.rs:65]: Ah, I was following AS3 here as well. I'll fix it in that case

**Comment by Ace4896** [comment]
Yeah, I don't have access to the Flash IDE - do you mind writing the test cases for this?

File: core/src/avm1/globals/error.rs (hunk 2 above)
Comment (Herschel) [reply core/src/avm1/globals/error.rs:24]: From what I can tell they are just standard values, so something like this in `create_proto`:
`error.define_value(gc_context, "message", "Error".into(), EnumSet::empty());`

By default, if you set no message, `error.message` just falls back to the prototype value and returns "Error".
But if you set `error.message = "foo";`, this creates a new value on the Error itself that takes precedence over the prototype value.

File: core/src/avm1/globals/error.rs (hunk 3 above)
Comment (Ace4896) [reply core/src/avm1/globals/error.rs:19]: Ah, I wasn't sure about what the type should've been since I wasn't able to run it - thanks for checking

**Comment by Ace4896** [comment]
@Herschel I think I've managed to address your comments, let me know if there's anything else I should change.

**Comment by Herschel** [comment]
Thank you!

**Pull request merged**